Support nested classes (#4178)

kamilrakoczy opened the pull request:
Fixes https://github.com/verilator/verilator/issues/4178.

This PR enable initial support for nested classes.

wsnyder commented on src/V3WidthCommit.cpp:
```diff
@@ -100,9 +100,15 @@ class WidthCommitVisitor final : public VNVisitor {
             const char* how = nullptr;
-            if (local && defClassp && refClassp != defClassp) {
+            // Inner nested classes can access `local` or `protected` members of their outer class
+            const auto nestedAccess = [refClassp](const AstClass*, const AstNode* memberp) {
+                return memberp == refClassp;
+            };
+            if (local && defClassp
+                && ((refClassp != defClassp) && !(defClassp->existsMember(nestedAccess)))) {
```
Do you have a negative test for this - that is make sure outer can't access local inner?

kamilrakoczy replied:
I've added it.

wsnyder reviewed (commented):
Well done, I especially appreciate the good testing.

In what ways is this only "initial" vs "complete"? (I'm surprised there weren't more back end and optimizer changes needed.)

kamilrakoczy commented:
> In what ways is this only "initial" vs "complete"?

I think, that it is quite complete. I've added tests for nested classes with parameters, extends, `new` function, access to member variables/functions, access to outer local/protected, static/virtual methods.

I added `initial` as I think there still might be corner cases, where nested classes might not work correctly.
